src/VisualStudio/Razor/RazorLanguageServiceClient.cs

```diff
@@ -29,7 +29,7 @@ public async Task<Session> CreateSessionAsync(Solution solution, object callback
                 return null;
             }
 
-            var innerSession = await _client.TryCreateSessionAsync(_serviceName, solution, callbackTarget, cancellationToken).ConfigureAwait(false);
+            var innerSession = await _client.TryCreateSessionAsync(_serviceName, solution, cancellationToken, callbackTarget).ConfigureAwait(false);
```

Comment:
why cancellationToken not last?

Reply:
It's not optional, `callbackTarget` is.

Reply:
pattern for that is two overloads (optionally one is an extension):

```
TryCreateSessionAsync(string name, object callback, CancellationToken cancellationToken);
TryCreateSessionAsync(string name, CancellationToken cancellationToken) => TryCreateSessionAsync(name, null, cancellationToken);
```

Reply:
This is simpler.

Reply:
Note that we have analyzer rule (CA1068) that checks for cancellation token being the last parameter and it explicitly allows optional parameters to follow.

Reply:
I think it's simpler to keep consistency with the rest of the IDE code.   the number of cases where cancellation is not last is enormously small.  i would prefer to not add to that.

Reply:
Why is it not good enough to follow rules that we recommend out customers to follow? Seems like since we have a rule and even an analyzer that's validating it, that should be what we do.

Reply:
Because it makes our codebase less consistent.  

>  that we recommend our customers to follow? 

There are lots of things that CA says that we do not follow.  Indeed, for a long time, CA was even banned from the codebase for being inappropriate.  

If we want to change our patterns/practices that should come with a dsicussion and decision that that's desirable, not just changing things mid PR.

> that should be what we do.

Reply:
> One thin the IDE tries to do is have people consistently pulling it in different directions. It's expensive and complex to maintain, and it happens continuously.

Well, that's a consequence of throwing the rules out and doing our own thing that we don't have analyzers to enforce.
This convention is not documented anywhere, so I didn't know we don't do this. If you say that's the convention then I can add it back, even though it adds extra clutter imo.

Reply:
BTW, we do have CA1068 enabled for Roslyn, so we do follow it.

Reply:
> Well, that's a consequence of throwing the rules out and doing our own thing that we don't have analyzers to enforce.

No it isn't.  Analyzers can identify thins that are not allowed.  But it doesn't address when the team only allows a subset of normally allowed cases for consistency.

> This convention is not documented anywhere, so I didn't know we don't do this.

It is the convention.  We haven't generally documented this though i would support such an effort.

> then I can add it back, even though it adds extra clutter imo.

I've fine with three approaches:

1. we get sign off that the new style is ok.  in which case, go ahead with this.
2. you use the existing style where CT should always be last.
3. you change the style, but we also update the rest of our code to follow this.

Basically, consistency is my prime goal here.  What i really dislike is jumping from area to area in the IDE and feeling that it's all following totally different rules depending on who wrote it and when.

Reply:
I have no problem with any of the above approaches. 

> the team only allows a subset of normally allowed cases for consistency.

I would rather avoid these, unless they can be expressed in .editorconfig file or are very specific to a particular Roslyn API usage.

I'd rather prefer if each stylistic rule was either agreed upon by the team and expressed in our .editorconfig file, or generally accepted as unimportant and thus not necessarily consistent across the code base. Of course, if a change is made to an existing class/file, the class/file obviously uses some style and that style is in the later category it should be preserved so that we minimize local inconsistencies.

BTW, re consistency - this change does not make anything inconsistent. The code base still follows the style as defined by CA1068 with or without my change.

Reply:
> I would rather avoid these, unless they can be expressed in .editorconfig file or are very specific to a particular Roslyn API usage.

I'd be fine with that.  however, as i have argued in the past, we need to *get* to that point.  i.e. the lack of enforcement tools now isn't a reason to be lax about enforcing these important consistency points.

Now that we have the infrastructure to find/enforce these style pieces, i would be 100% behind efforts to actually codify this stuff.  I certainly know it would help and would serve to disseminate information. 

> CA1068 

This isn't something meaningful to the code.  the CA rules have never been ones that hte team has chosen to be the ones that are followed.  StyleCop and the like are a set of tools that can bring some consistency, but haven't been the consistency chosen by the team for something like 15 years at this point.  

If the team wants to move entirely to CA, that's fine.  But that needs to be a decision made formally.  My preference though would be to not go that direction and instead codify the rules that the IDE has been built off of.  It would also be good as an appropriate form of dogfooding of our own tools/config systems.

Reply:
TLDR: Feel free to start adding enforcement rules to support IDE coding patterns.  I'm 100% behind that.  I'd contribute to such an effort.

Reply:
OK. So maybe a good start would be to write up the "consistency" rules we want to be enforced. I don't know what those should be. If you know of any feel free to file an issue that lists them with the goal of having them checked.

Re CA rules - I don't know why would they not be meaningful to our code. Regardless, we have  a subset of CA rules enabled today and those are being enforced.

Reply:
If you want to open an isuse i can definitely dump some that i know about that we see a fair amount.

> I don't know why would they not be meaningful to our code.

Let me rephrase:

They're meaningful insofar that we don't allow violations of them.  However, they're not to be taken as approval if they all pass.  Similar to the compiler layer, it's totally possible to pass all the automated tooling but still be unacceptable as per the informal rules and styles that the team has come to expect over hte last 20 years :)

In other words: they're a starting point.  They're not the ending point.

Reply:
We discussed this topic at IDE design meeting today. The conclusions are:
1) Follow CA1068 as is, i.e. allow cancellation token to precede optional parameters.
Jason vaguely remembers that we used to have the convention for CT to be always last, but nobody was sure and no one objected with changing it either.
2) File issues on conventions that can be enforced by analyzers, we want them to be enforced, but are not enforced today.

Reply:
SGTM.  Thanks for getting team buyoff!